pkg/reconciler/managed/changelogger.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2024 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package managed
+
+import (
+	"context"
+	"fmt"
+
+	changelogs "github.com/crossplane/crossplane-runtime/apis/changelogs/proto/v1alpha1"
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/reference"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+// changeLogger processes changes to resources and helps to send them to the
+// change log service
+type changeLogger struct {
+	enabled         bool
+	client          changelogs.ChangeLogServiceClient
+	providerVersion string
+	log             logging.Logger
+	record          event.Recorder
+	snapshot        *structpb.Struct
+}
+
+// newChangeLogger creates a new changeLogger initialized with the given values
+func newChangeLogger(enabled bool, client changelogs.ChangeLogServiceClient, log logging.Logger, record event.Recorder, providerVersion string) *changeLogger {
+	return &changeLogger{
+		enabled:         enabled,
+		client:          client,
+		providerVersion: providerVersion,
+		log:             log,
+		record:          record,
+	}
+}
+
+// takeSnapshot attempts to take a snapshot of the given managed resource by
+// converting it to a Struct that will be sent to the change logs after an
+// external operation is performed, if the change logs feature is enabled. If
+// there is an error during this process, an event will be recorded and no
+// snapshot will be saved.
+func (c *changeLogger) takeSnapshot(managed resource.Managed) {
```

Comment:
I wonder whether this type really needs three methods (snapshot, new entry, record entry).

If I was designing this I'd probably:

1. Take a deep copy of `managed` where you currently take the snapshot
2. Pass the deep copy, opType, error and additional details to `recordChangeLog` and have it form the entry

Reply:
Yep, I like this approach as done in https://github.com/crossplane/crossplane-runtime/pull/754/commits/c23f09b8314e5a7d11e0438243572a0b38cfaff1